### PR TITLE
Conversation page

### PR DESCRIPTION
**Pull request by whsieh**
Implemented the conversation page. Until the authentication patch comes in, user authentication is stubbed out.

### CODE REVIEW

File: features/step_definitions/app_steps.rb [hunk 1]
```diff
@@ -95,3 +98,49 @@
         raise "No check for the #{page_name} page."
     end
 end
+
+# HACK This is probably worthy of some award. Not the good kind.
+And /^I should see the latest post$/ do
+    sleep 2
```
Comment (whsieh) [inline features/step_definitions/app_steps.rb:94]: The `sleep 2` is arbitrary, but on my machine this test often fails if I don't wait for a least this amount of time.

File: app/assets/javascripts/homepage/conversation.coffee [hunk 2]
```diff
@@ -0,0 +1,97 @@
+conversation = angular.module("Conversation", ["ngRoute", "SharedServices"])
+
+conversation.config(["$routeProvider", ($routeProvider) ->
+    $routeProvider.when("/conversation/:id", {
+        templateUrl: "/assets/conversation.html",
+        controller: "ConversationController"
+    })
+
+]).directive("ngEscape", ->
```
Comment (leohtkam) [inline app/assets/javascripts/homepage/conversation.coffee]: Don't prefix a directive with "ng" if it's not from angular. Probably prefix with "my" or something.

File: app/assets/javascripts/homepage/conversation.coffee [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    postSubmissionInProgress = false
+    conversation = {}
+    postsById = {}
+    currentPostEditId = null
+    postSubmissionTimer = null
+    # Scope methods and models
+    $scope.editPostText = ""
+    $scope.postSubmitError = ""
+    $scope.timeElapsedMessage = (timestamp) -> TimeUtil.timeIntervalAsString(TimeUtil.timeSince1970InSeconds() - TimeUtil.timeFromTimestampInSeconds(timestamp)) + " ago"
+    $scope.postEditorClass = -> if currentlyEditingPost then "" else "hidden"
```
Comment (leohtkam) [inline app/assets/javascripts/homepage/conversation.coffee]: Why not use ngHide if the hidden class is just setting display to none.

File: app/assets/javascripts/homepage/conversation.coffee [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    $scope.postLengthClass = -> if currentlyEditingPost then "posts-shortened" else "posts-full-length"
+    $scope.addPostClicked = ->
+        $scope.postSubmitError = ""
+        currentlyEditingPost = true
+        $scope.editPostText = "" if currentPostEditId # Clear the editor if the user was previously editing an existing post.
```
Comment (leohtkam) [inline app/assets/javascripts/homepage/conversation.coffee:39]: Just a comment on style, but why not check currentlyEditingPost instead?

File: app/assets/javascripts/homepage/conversation.coffee [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        currentPostEditId = null
+        dispatchFocusTextarea()
+
+    $scope.cancelPostClicked = -> currentlyEditingPost = false
```
Comment (leohtkam) [inline app/assets/javascripts/homepage/conversation.coffee]: A comment for all the somethingClicked methods: might as well call them somethingButtonClicked to make it more clear.

File: app/assets/javascripts/homepage/conversation.coffee (hunk 4 above)
Comment (whsieh) [inline app/assets/javascripts/homepage/conversation.coffee:39]: for line 37? That's not what I'm trying to check.

File: app/assets/javascripts/homepage/conversation.coffee (hunk 5 above)
Comment (whsieh) [inline app/assets/javascripts/homepage/conversation.coffee]: They're not buttons, they're links. somethingButtonClicked would be misleading. somethingClicked is also more concise.

File: app/assets/javascripts/homepage/conversation.coffee (hunk 3 above)
Comment (whsieh) [inline app/assets/javascripts/homepage/conversation.coffee]: Good point. Will fix.

File: app/assets/javascripts/homepage/conversation.coffee (hunk 4 above)
Comment (leohtkam) [inline app/assets/javascripts/homepage/conversation.coffee:39]: I see what you mean. I fail at reading...

File: features/step_definitions/app_steps.rb [hunk 6]
```diff
@@ -48,6 +48,9 @@
     arena = users[0].arenas.create :user1 => users[0], :user2 => users[1]
     conversation = arena.conversations.create :title => "What messages are you posting?"
     messages.each_with_index do |message, index|
+        if message == "(and a huge post)"
+            message = "bacon strips & " * 40
+        end
```
Comment (leohtkam) [inline features/step_definitions/app_steps.rb:26]: lolz

**Comment by leohtkam** [comment]
+1

File: app/assets/javascripts/homePage.js [hunk 7]
```diff
@@ -4,5 +4,6 @@
 //= require_tree ./shared
 //= require_tree ./sidebar
 //= require_tree ./profile
+//= require_tree ./homepage
 
-app = angular.module("HomePage", ["ui.bootstrap", "ngRoute", "Profile", "Sidebar"])
+app = angular.module("HomePage", ["ui.bootstrap", "ngRoute", "SharedServices", "Profile", "Sidebar", "Conversation"])
```
Comment (nickfong) [inline app/assets/javascripts/homePage.js]: `SharedServices` and the `//= require_tree ./homepage` were removed during refactoring in 46d2683.  Are they necessary here?

**Comment by nickfong** [comment]
+1 once my comment re. `homePage.js` is resolved.

File: app/assets/javascripts/homePage.js (hunk 7 above)
Comment (whsieh) [inline app/assets/javascripts/homePage.js]: The SharedServices has been removed. However, the require_tree ./homepage is necessary to include the conversation script.

**Comment by nickfong** [comment]
Revision looks good.  +1

**Comment by leohtkam** [comment]
+1 for the polling.

**Pull request merged**